api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java

```diff
@@ -0,0 +1,324 @@
+import org.projectnessie.client.http.Status;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A simple OAuth2 client that supports the "client_credentials" grant type for fetching new access
+ * tokens, using pre-defined client credentials.
+ *
+ * <p>This client also supports refreshing access tokens using both the refresh token method defined
+ * in RFC 6749 and the token exchange method defined in RFC 8693.
+ *
+ * <p>If you don't need to refresh access tokens, you can use the {@link
+ * org.projectnessie.client.auth.BearerAuthenticationProvider BearerAuthenticationProvider} instead
+ * and provide the token to use directly through configuration.
+ */
+class OAuth2Client implements OAuth2Authenticator, Closeable {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(OAuth2Client.class);
+
+  private final String grantType;
+  private final String username;
+  private final byte[] password;
+  private final String scope;
+  private final Duration defaultAccessTokenLifespan;
+  private final Duration defaultRefreshTokenLifespan;
+  private final Duration refreshSafetyWindow;
+  private final boolean tokenExchangeEnabled;
+  private final HttpClient httpClient;
+  private final ScheduledExecutorService executor;
+  private final boolean shouldCloseExecutor;
+  private final ObjectMapper objectMapper;
+  private final CompletableFuture<Void> started = new CompletableFuture<>();
+
+  private volatile CompletionStage<Tokens> currentTokensStage;
+  private volatile ScheduledFuture<?> tokenRefreshFuture;
+
+  OAuth2Client(OAuth2ClientParams params) {
+    grantType = params.getGrantType();
+    username = params.getUsername().orElse(null);
+    password = params.getPassword().orElse(null);
+    scope = params.getScope().orElse(null);
+    defaultAccessTokenLifespan = params.getDefaultAccessTokenLifespan();
+    defaultRefreshTokenLifespan = params.getDefaultRefreshTokenLifespan();
+    refreshSafetyWindow = params.getRefreshSafetyWindow();
+    tokenExchangeEnabled = params.getTokenExchangeEnabled();
+    httpClient = params.getHttpClient().addResponseFilter(this::checkErrorResponse).build();
+    executor = params.getExecutor().orElseGet(OAuth2Client::createDefaultExecutor);
+    shouldCloseExecutor = !params.getExecutor().isPresent();
+    objectMapper = params.getObjectMapper();
+    currentTokensStage = started.thenApplyAsync((v) -> fetchNewTokens(), executor);
+    currentTokensStage.thenAccept(this::scheduleTokensRenewal);
+  }
+
+  private static ScheduledExecutorService createDefaultExecutor() {
+    return Executors.newSingleThreadScheduledExecutor(
+        runnable -> {
+          Thread thread = Executors.defaultThreadFactory().newThread(runnable);
+          thread.setDaemon(true);
+          thread.setName("oauth2-client");
+          return thread;
+        });
+  }
+
+  @Override
+  public AccessToken authenticate() {
+    return getCurrentTokens().getAccessToken();
```

Comment:
Would it be a lot of effort to have this covered in a unit test? Does it make sense?
(I see, it's covered in the IT, so not really an issue.)

Reply:
That should be possible with some mocking. Good idea.

Reply:
Done in 91b5ca7c6.